## 파이썬 300제(151번 ~ 300번)
- [파이썬 300제](https://wikidocs.net/78768)

### #151. 리스트에는 네 개의 정수가 저장돼 있다.

> 리스트 = [3, -20, -3, 44]
### for문을 사용해서 리스트의 음수를 출력하라.

>-20  
-3

In [1]:
리스트 = [3, -20, -3, 44]
for i in 리스트:
    if i < 0:
        print(i)

-20
-3


### #152. for문을 사용해서 3의 배수만을 출력하라.

> 리스트 = [3, 100, 23, 44]

In [2]:
리스트 = [3, 100, 23, 44]
for i in 리스트:
    if i % 3 == 0:
        print(i)

3


### #153. 리스트에서 20 보다 작은 3의 배수를 출력하라

> 리스트 = [13, 21, 12, 14, 30, 18]

In [3]:
리스트 = [13, 21, 12, 14, 30, 18]
for i in 리스트:
    if i < 20 and i % 3 == 0:
        print(i)

12
18


### #154. 리스트에서 세 글자 이상의 문자를 화면에 출력하라

> 리스트 = ["I", "study", "python", "language", "!"]

In [4]:
리스트 = ["I", "study", "python", "language", "!"]
for i in 리스트:
    if len(i) >= 3:
        print(i)

study
python
language


### #155. 리스트에서 대문자만 화면에 출력하라.

> 리스트 = ["A", "b", "c", "D"]

In [5]:
리스트 = ["A", "b", "c", "D"]
for i in 리스트:
    if i.isupper() == True:
        print(i)

A
D


### #156. 리스트에서 소문자만 화면에 출력하라.

> 리스트 = ["A", "b", "c", "D"]

In [6]:
리스트 = ["A", "b", "c", "D"]
for i in 리스트:
    if i.isupper() == False:
        print(i)

b
c


### #157. 이름의 첫 글자를 대문자로 변경해서 출력하라.

> 리스트 = ['dog', 'cat', 'parrot']

In [9]:
리스트 = ['dog', 'cat', 'parrot']
for i in 리스트:
    print(i[0].upper() + i[1:])

Dog
Cat
Parrot


### #158. 파일 이름이 저장된 리스트에서 확장자를 제거하고 파일 이름만 화면에 출력하라. (힌트: split() 메서드)

> 리스트 = ['hello.py', 'ex01.py', 'intro.hwp']

In [11]:
리스트 = ['hello.py', 'ex01.py', 'intro.hwp']
for i in 리스트:
    print(i.split(".")[0])

hello
ex01
intro


### #159. 파일 이름이 저장된 리스트에서 확장자가 .h인 파일 이름을 출력하라.

> 리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']

In [12]:
리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']
for i in 리스트:
    if i.split(".")[1] == "h":
        print(i)

intra.h
define.h


### #160. 파일 이름이 저장된 리스트에서 확장자가 .h나 .c인 파일을 화면에 출력하라.

> 리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']

In [13]:
리스트 = ['intra.h', 'intra.c', 'define.h', 'run.py']
for i in 리스트:
    split = i.split(".")
    if split[1] == "h" or split[1] == "c":
        print(i)

intra.h
intra.c
define.h


### #161. for문과 range 구문을 사용해서 0~99까지 한 라인에 하나씩 순차적으로 출력하는 프로그램을 작성하라.

In [14]:
for i in range(100):
    print(i)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99


### #162. 월드컵은 4년에 한 번 개최된다. range()를 사용하여 2002~2050년까지 중 월드컵이 개최되는 연도를 출력하라.

In [15]:
for i in range(2002, 2051, 4):
    print(i)

2002
2006
2010
2014
2018
2022
2026
2030
2034
2038
2042
2046
2050


### #163. 1부터 30까지의 숫자 중 3의 배수를 출력하라.

In [18]:
for i in range(3, 31, 3):
    print(i)

3
6
9
12
15
18
21
24
27
30


### #164. 99부터 0까지 1씩 감소하는 숫자들을, 한 라인에 하나씩 출력하라.

In [21]:
for i in range(100):
    print(99 - i)

99
98
97
96
95
94
93
92
91
90
89
88
87
86
85
84
83
82
81
80
79
78
77
76
75
74
73
72
71
70
69
68
67
66
65
64
63
62
61
60
59
58
57
56
55
54
53
52
51
50
49
48
47
46
45
44
43
42
41
40
39
38
37
36
35
34
33
32
31
30
29
28
27
26
25
24
23
22
21
20
19
18
17
16
15
14
13
12
11
10
9
8
7
6
5
4
3
2
1
0


### #165. for문을 사용해서 아래와 같이 출력하라.

> 0.0  
0.1  
0.2  
0.3  
0.4  
0.5  
...  
0.9  

In [23]:
for i in range(10):
    print(i / 10)

0.0
0.1
0.2
0.3
0.4
0.5
0.6
0.7
0.8
0.9


### #166. 구구단 3단을 출력하라.

In [24]:
for i in range(1, 10):
    print("3 x", i, "=", 3 * i)

3 x 1 = 3
3 x 2 = 6
3 x 3 = 9
3 x 4 = 12
3 x 5 = 15
3 x 6 = 18
3 x 7 = 21
3 x 8 = 24
3 x 9 = 27


### #167. 구구단 3단을 출력하라. 단 홀수 번째만 출력한다.

In [25]:
for i in range(1, 10, 2):
    print("3 x", i, "=", 3 * i)

3 x 1 = 3
3 x 3 = 9
3 x 5 = 15
3 x 7 = 21
3 x 9 = 27


### #168. 1~10까지의 숫자에 대해 모두 더한 값을 출력하는 프로그램을 for 문을 사용하여 작성하라.

In [27]:
k = 0
for i in range(1, 11):
    k += i
print("합 :", k)

합 : 55


### #169. 1~10까지의 숫자 중 모든 홀수의 합을 출력하는 프로그램을 for 문을 사용하여 작성하라.

In [28]:
k = 0
for i in range(1, 11, 2):
    k += i
print("합 :", k)

합 : 25


### #170. 1~10까지의 숫자를 모두 곱한 값을 출력하는 프로그램을 for 문을 사용하여 작성하라.

In [29]:
k = 1
for i in range(1, 11):
    k *= i
print(k)

3628800


### #171. 아래와 같이 리스트의 데이터를 출력하라. 단, for문과 range문을 사용하라.

> price_list = [32100, 32150, 32000, 32500]

>32100  
32150  
32000  
32500  

In [30]:
price_list = [32100, 32150, 32000, 32500]
for i in range(4):
    print(price_list[i])

32100
32150
32000
32500


### #172. 아래와 같이 리스트의 데이터를 출력하라. 단, for문과 range문을 사용하라.

> price_list = [32100, 32150, 32000, 32500]

>0 32100  
1 32150  
2 32000  
3 32500

In [31]:
price_list = [32100, 32150, 32000, 32500]
for i in range(4):
    print(i, price_list[i])

0 32100
1 32150
2 32000
3 32500


### #173. 아래와 같이 리스트의 데이터를 출력하라. 단, for문과 range문을 사용하라.

> price_list = [32100, 32150, 32000, 32500]

>3 32100  
2 32150  
1 32000  
0 32500  

In [32]:
price_list = [32100, 32150, 32000, 32500]
for i in range(4):                # 4라는 숫자 대신 len(price_list)을 사용하는 것이 더 좋다.
    print(3 - i, price_list[i])

3 32100
2 32150
1 32000
0 32500


### #174. 아래와 같이 리스트의 데이터를 출력하라. 단, for문과 range문을 사용하라.

> price_list = [32100, 32150, 32000, 32500]

>100 32150  
110 32000  
120 32500

In [34]:
price_list = [32100, 32150, 32000, 32500]
for i in range(1, 4):
    print(100 + 10 * (i - 1), price_list[i])

100 32150
110 32000
120 32500


### #175. my_list를 아래와 같이 출력하라.

> my_list = ["가", "나", "다", "라"]

>가 나  
나 다  
다 라

In [35]:
my_list = ["가", "나", "다", "라"]
for i in range(3):
    print(my_list[i], my_list[i + 1])

가 나
나 다
다 라


### #176. 리스트를 아래와 같이 출력하라.

> my_list = ["가", "나", "다", "라", "마"]

>가 나 다  
나 다 라  
다 라 마

In [37]:
my_list = ["가", "나", "다", "라", "마"]
for i in range(3):
    print(my_list[i], my_list[i + 1], my_list[i + 2])

가 나 다
나 다 라
다 라 마


### #177. 반복문과 range 함수를 사용해서 my_list를 아래와 같이 출력하라.

>my_list = ["가", "나", "다", "라"]

>라 다  
다 나  
나 가

In [38]:
my_list = ["가", "나", "다", "라"]
for i in range(3):
    print(my_list[3 - i], my_list[2 - i])

라 다
다 나
나 가


### #178. 리스트에는 네 개의 정수가 저장되어 있다. 각각의 데이터에 대해서 자신과 우측값과의 차분값을 화면에 출력하라.

> my_list = [100, 200, 400, 800]
### 예를들어 100을 기준으로 우측에 위치한 200과의 차분 값를 화면에 출력하고, 200을 기준으로 우측에 위치한 400과의 차분값을 화면에 출력한다. 이어서 400을 기준으로 우측에 위치한 800과의 차분값을 화면에 출력한다.

> 100  
200  
400  

In [39]:
my_list = [100, 200, 400, 800]
for i in range(3):
    print(my_list[i + 1] - my_list[i])

100
200
400


### #179. 리스트에는 6일 간의 종가 데이터가 저장되어 있다. 종가 데이터의 3일 이동 평균을 계산하고 이를 화면에 출력하라.

> my_list = [100, 200, 400, 800, 1000, 1300]
### 첫 번째 줄에는 100, 200, 400의 평균값이 출력된다. 두 번째 줄에는 200, 400, 800의 평균값이 출력된다. 같은 방식으로 나머지 데이터의 평균을 출력한다.

>233.33333333333334  
466.6666666666667  
733.3333333333334  
1033.3333333333333

In [40]:
my_list = [100, 200, 400, 800, 1000, 1300]
for i in range(4):
    print((my_list[i] + my_list[i + 1] + my_list[i + 2]) / 3)

233.33333333333334
466.6666666666667
733.3333333333334
1033.3333333333333


### #180. 리스트에 5일간의 저가, 고가 정보가 저장돼 있다. 고가와 저가의 차를 변동폭이라고 정의할 때, low, high 두 개의 리스트를 사용해서 5일간의 변동폭을 volatility 리스트에 저장하라.

>low_prices  = [100, 200, 400, 800, 1000]  
high_prices = [150, 300, 430, 880, 1000]

In [41]:
low_prices = [100, 200, 400, 800, 1000]
high_prices = [150, 300, 430, 880, 1000]

volatility = []
for i in range(5):
    k = high_prices[i] - low_prices[i]
    volatility.append(k)
print(volatility)

[50, 100, 30, 80, 0]


### #181. 아래 표에서 하나의 행을 하나의 리스트로, 총 3개의 리스트를 갖는 이차원 리스트 apart를 정의하라.

>101호	102호  
>201호	202호  
>301호	302호

In [1]:
apart = [["101호", "102호"], ["201호", '202호'], ["301호", "302호"]]
apart

[['101호', '102호'], ['201호', '202호'], ['301호', '302호']]

### #182. 아래 표에서 하나의 열을 하나의 리스트로, 총 2개의 리스트를 갖는 이차원 리스트 stock을 정의하라.

>시가	종가  
100	80  
200	210  
300	330  

In [2]:
stock = [["시가", "100", "200", "300"], ["종가", "80", "210", "330"]]
stock

[['시가', '100', '200', '300'], ['종가', '80', '210', '330']]

### #183. 아래 표를 stock 이름의 딕셔너리로 표현하라.시가를 key로 저장하고, 나머지 같은 열의 데이터를 리스트로 저장해서 value로 저장한다. 종가 역시 key로 저장하고 나머지 같은 열의 데이터를 리스트로 저장해서 value로 저장한다.

>시가	종가  
100	80  
200	210  
300	330  

In [4]:
stock = {"시가" : [100, 200, 300], "종가" : [80, 210, 330]}

### #184. 아래 표를 stock 이라는 이름의 딕셔너리로 표현하라. 날짜를 key로 저장하고, 나머지 같은 행의 데이터를 리스트로 저장해서 value로 저장한다. 첫 열이 날짜이다.

>10/10	80	110	70	90  
10/11	210	230	190	200  

In [3]:
stock = {"10/10" : [80, 110, 70, 90], "10/11" : [210, 230, 190, 200]}

### #185. 리스트에 저장된 데이터를 아래와 같이 출력하라.

>apart = [ [101, 102], [201, 202], [301, 302] ]

>101 호  
102 호  
201 호  
202 호  
301 호  
302 호

In [7]:
apart = [ [101, 102], [201, 202], [301, 302] ]
for i in apart:
    print(i[0], "호")
    print(i[1], "호")

101 호
102 호
201 호
202 호
301 호
302 호


In [9]:
# 정답 - 이중 for문을 사용해서 나타낼 수도 있다.
for row in apart:
    for col in row:
        print(col, "호")

101 호
102 호
201 호
202 호
301 호
302 호


### #186. 리스트에 저장된 데이터를 아래와 같이 출력하라.

> apart = [ [101, 102], [201, 202], [301, 302] ]

>301 호  
302 호  
201 호  
202 호  
101 호  
102 호

In [10]:
apart = [ [101, 102], [201, 202], [301, 302] ]

for i in apart[::-1]:
    for k in i:
        print(k, "호")

301 호
302 호
201 호
202 호
101 호
102 호


### #187. 리스트에 저장된 데이터를 아래와 같이 출력하라.

> apart = [ [101, 102], [201, 202], [301, 302] ]

>302 호  
301 호  
202 호  
201 호  
102 호  
101 호

In [11]:
apart = [ [101, 102], [201, 202], [301, 302] ]

for i in apart[::-1]:
    for k in i[::-1]:
        print(k, "호")

302 호
301 호
202 호
201 호
102 호
101 호


### #188. 리스트에 저장된 데이터를 아래와 같이 출력하라.

> apart = [ [101, 102], [201, 202], [301, 302] ]

>101 호  
\-----  
102 호  
\-----  
201 호  
\-----  
202 호  
\-----  
301 호  
\-----  
302 호  
\-----  

In [13]:
apart = [ [101, 102], [201, 202], [301, 302] ]

for i in apart:
    for k in i:
        print(k, "호")
        print("-----")

101 호
-----
102 호
-----
201 호
-----
202 호
-----
301 호
-----
302 호
-----


### #189. 리스트에 저장된 데이터를 아래와 같이 출력하라.

> apart = [ [101, 102], [201, 202], [301, 302] ]

> 101 호  
102 호  
\-----  
201 호  
202 호  
\-----  
301 호  
302 호  
\-----  

In [14]:
apart = [ [101, 102], [201, 202], [301, 302] ]

for i in apart:
    for k in i:
        print(k, "호")
    print("-----")

101 호
102 호
-----
201 호
202 호
-----
301 호
302 호
-----


### #190. 리스트에 저장된 데이터를 아래와 같이 출력하라.

> apart = [ [101, 102], [201, 202], [301, 302] ]

>101 호  
102 호  
201 호  
202 호  
301 호  
302 호  
\-----


In [15]:
apart = [ [101, 102], [201, 202], [301, 302] ]

for i in apart:
    for k in i:
        print(k, "호")
print("-----")

101 호
102 호
201 호
202 호
301 호
302 호
-----


### #191. data에는 매수한 종목들의 OHLC (open/high/low/close) 가격 정보가 바인딩 되어있다.

>data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]

### 수수료를 0.014 %로 가정할 때, 각 가격에 수수료를 포함한 가격을 한라인에 하나씩 출력하라.

In [17]:
data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]

for i in data:
    for j in i:
        print(j * 1.00014)

2000.28
3050.427
2050.2870000000003
1980.2772
7501.05
2050.2870000000003
2050.2870000000003
1980.2772
15452.163
15052.107
15552.177
14902.086000000001


### #192. 191번의 출력 결과에 행단위로 "----" 구분자를 추가하라.

In [18]:
data = [
    [ 2000,  3050,  2050,  1980],
    [ 7500,  2050,  2050,  1980],
    [15450, 15050, 15550, 14900]
]

for i in data:
    for j in i:
        print(j * 1.00014)
    print("----")

2000.28
3050.427
2050.2870000000003
1980.2772
----
7501.05
2050.2870000000003
2050.2870000000003
1980.2772
----
15452.163
15052.107
15552.177
14902.086000000001
----


### #193. 191 번 문제의 결괏값을 result 이름의 리스트에 1차원 배열로 저장하라.

In [20]:
result = []
for i in data:
    for j in i:
        result.append(j * 1.00014)
        
print(result)

[2000.28, 3050.427, 2050.2870000000003, 1980.2772, 7501.05, 2050.2870000000003, 2050.2870000000003, 1980.2772, 15452.163, 15052.107, 15552.177, 14902.086000000001]


### #194. 191번 문제의 결괏값을 result 이름의 리스트에 2차원 배열로 저장하라. 각 행에 대한 데이터끼리 리스트에 저장되어야 한다.

In [26]:
result = []

for i in data:
    a = []
    for j in i:
        a.append(j * 1.00014)
    result.append(a)

print(result)

[[2000.28, 3050.427, 2050.2870000000003, 1980.2772], [7501.05, 2050.2870000000003, 2050.2870000000003, 1980.2772], [15452.163, 15052.107, 15552.177, 14902.086000000001]]


### #195. ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 화면에 종가데이터를 출력하라.

> ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

In [27]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

for i in ohlc[1:]:
    print(i[-1])

100
190
310


### #196. ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 종가가 150원보다 큰경우에만 종가를 출력하라.

> ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

In [28]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

for i in ohlc[1:]:
    if i[-1] > 150:
        print(i[-1])

190
310


### #197. ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 종가가 시가 보다 크거나 같은 경우에만 종가를 출력하라.

> ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

In [30]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

for i in ohlc[1:]:
    if i[-1] >= i[0]:
        print(i[-1])

100
310


### #198. ohlc 리스트에는 시가(open), 고가 (high), 저가 (low) , 종가(close)가 날짜별로 저장돼 있다. 고가와 저가의 차이를 변동폭으로 정의할 때 변동폭을 volatility 이름의 리스트에 저장하라.

> ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

In [32]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

volatility = []
for i in ohlc[1:]:
    volatility.append(i[1] - i[2])
print(volatility)

[40, 30, 10]


### #199. 리스트에는 3일 간의 ohlc 데이터가 저장돼 있다. 종가가 시가보다 높은 날의 변동성 (고가 - 저가)을 화면에 출력하라.

> ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

In [33]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

for i in ohlc[1:]:
    if i[3] > i[0]:
        print(i[3] - i[0])

10


### #200. 리스트에는 3일 간의 ohlc 데이터가 저장돼 있다. 시가에 매수해서 종가에 매도 했을 경우 총 수익금을 계산하라.

> ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

In [36]:
ohlc = [["open", "high", "low", "close"],
        [100, 110, 70, 100],
        [200, 210, 180, 190],
        [300, 310, 300, 310]]

profit = 0
for i in ohlc[1:]:
    profit += (i[3] - i[0])
print(profit)

0


### #201. "비트코인" 문자열을 화면에 출력하는 print_coin() 함수를 정의하라. 

In [38]:
def print_coin():
    print("비트코인")

### #202. 201번에서 정의한 함수를 호출하라.

In [39]:
print_coin()

비트코인


### #203. 201번에서 정의한 print_coin 함수를 100번호출하라.

In [41]:
for i in range(100):
    print_coin()

비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인
비트코인


### #204. "비트코인" 문자열을 100번 화면에 출력하는 print_coins() 함수를 정의하라.

In [42]:
def print_coins():
    for i in range(100):
        print("비트코인")

### #205. 아래의 에러가 발생하는 이유에 대해 설명하라.

In [43]:
hello()
def hello():
    print("Hi")

# hello()라는 함수가 정의되기 전 hello()라는 함수를 호출하였기 때문에 에러가 발생한다.

NameError: name 'hello' is not defined

### #206. 아래 코드의 실행 결과를 예측하라.

In [44]:
def message() :
    print("A")
    print("B")

message()
print("C")
message()

# A
# B
# C
# A
# B

A
B
C
A
B


### #207. 아래 코드의 실행 결과를 예측하라.

In [45]:
print("A")

def message() :
    print("B")

print("C")
message()

# A
# C
# B

A
C
B


### #208. 아래 코드의 실행 결과를 예측하라.

In [46]:
print("A")
def message1() :
    print("B")
print("C")
def message2() :
    print("D")
message1()
print("E")
message2()

# A
# C
# B
# E
# D

A
C
B
E
D


### #209. 아래 코드의 실행 결과를 예측하라.

In [47]:
def message1():
    print("A")

def message2():
    print("B")
    message1()

message2()

# B
# A

B
A


### #210. 아래 코드의 실행 결과를 예측하라.

In [48]:
def message1():
    print("A")

def message2():
    print("B")

def message3():
    for i in range (3) :
        message2()
        print("C")
    message1()

message3()

# B
# C
# B
# C
# B
# C
# A

B
C
B
C
B
C
A


### #211. 함수의 호출 결과를 예측하라.

In [49]:
def 함수(문자열) :
    print(문자열)

함수("안녕")
함수("Hi")

# 안녕
# Hi

안녕
Hi


### #212. 함수의 호출 결과를 예측하라.

In [50]:
def 함수(a, b) :
    print(a + b)

함수(3, 4)
함수(7, 8)

# 7
# 15

7
15


### #213. 아래와 같은 에러가 발생하는 원인을 설명하라.

In [51]:
def 함수(문자열) :
    print(문자열)
    
함수()
# 함수()라는 함수는 함수를 호출할 때 1개의 파라미터를 입력해야 하는데 그렇지 않았기 때문에 에러가 발생한다.

TypeError: 함수() missing 1 required positional argument: '문자열'

### #214. 아래와 같은 에러가 발생하는 원인을 설명하라.

In [52]:
def 함수(a, b) :
    print(a + b)

함수("안녕", 3)
# a와 b에 파라미터를 입력할 때 문자열(str)은 문자열끼리, 정수(int)는 정수끼리 등 같은 타입을 입력해야하는데 str과 int를 입력해주었기 때문이다.

TypeError: can only concatenate str (not "int") to str

### #215. 하나의 문자를 입력받아 문자열 끝에 ":D" 스마일 문자열을 이어 붙여 출력하는 print_with_smile 함수를 정의하라.

In [53]:
def print_with_smile(문자열):
    print(문자열 + ":D")

### #216. 215에서 정의한 함수를 호출하라. 파라미터는 "안녕하세요"로 입력하라.

In [55]:
print_with_smile("안녕하세요")

안녕하세요:D


### #217. 현재 가격을 입력 받아 상한가 (30%)를 출력하는 print_upper_price 함수를 정의하라.

In [56]:
def print_upper_price(price):
    print(price * 1.3)

### #218. 두 개의 숫자를 입력받아 두 수의 합을 출력하는 print_sum 함수를 정의하라.

In [57]:
def print_sum(a, b):
    print(a + b)

### #219. 두 개의 숫자를 입력받아 합/차/곱/나눗셈을 출력하는 print_arithmetic_operation 함수를 작성하라.

In [58]:
def print_arithmetic_operation(a, b):
    print(a, "+", b, "=", a + b)
    print(a, "-", b, "=", a - b)
    print(a, "*", b, "=", a * b)
    print(a, "/", b, "=", a / b)
    
print_arithmetic_operation(3, 4)

3 + 4 = 7
3 - 4 = -1
3 * 4 = 12
3 / 4 = 0.75


### #220. 세 개의 숫자를 입력받아 가장 큰수를 출력하는 print_max 함수를 정의하라. 단 if 문을 사용해서 수를 비교하라.

In [59]:
def print_max(a, b, c):
    if a >= b and a >= c:
        print(a)
    elif b >= a and b >= c:
        print(b)
    else:
        print(c)

In [60]:
# 정답 - 0과 a 비교해서 큰 수가 max_val. 다시 max_val과 b를 비교해서 큰 수가 max_val. 다시 max_val과 c를 비교해서 큰 수가 max_val.
# max() 내장 함수도 비슷한 원리로 작동한다.
def print_max(a, b, c) :
    max_val = 0
    if a > max_val :
        max_val = a
    if b > max_val :
        max_val = b
    if c > max_val :
        max_val = c
    print(max_val)

### #221. 입력된 문자열을 역순으로 출력하는 print_reverse 함수를 정의하라.

In [62]:
def print_reverse(str):
    print(str[::-1])

### #222. 성적 리스트를 입력 받아 평균을 출력하는 print_score 함수를 정의하라.

In [63]:
def print_score(score_list):
    print(sum(score_list) / len(score_list))

### #223. 하나의 리스트를 입력받아 짝수만 화면에 출력하는 print_even 함수를 정의하라.

In [66]:
def print_even(my_list):
    for i in my_list:
        if i % 2 == 0:
            print(i)

### #224. 하나의 딕셔너리를 입력받아 딕셔너리의 key 값을 화면에 출력하는 print_keys 함수를 정의하라.

In [67]:
def print_keys(dic):
    for k in dic.keys():
        print(k)

### #225. my_dict에는 날짜를 키값으로 OHLC가 리스트로 저장돼 있다.

> my_dict = {"10/26" : [100, 130, 100, 100],
           "10/27" : [10, 12, 10, 11]}
### my_dict와 날짜 키값을 입력받아 OHLC 리스트를 출력하는 print_value_by_key 함수를 정의하라.

In [69]:
def print_value_by_key(my_dict, key):
    print(my_dict[key])
    
my_dict = {"10/26" : [100, 130, 100, 100],
           "10/27" : [10, 12, 10, 11]}
print_value_by_key (my_dict, "10/26")

[100, 130, 100, 100]


### #226. 입력 문자열을 한 줄에 다섯글자씩 출력하는 print_5xn(string) 함수를 작성하라.

In [73]:
def print_5xn(string):
    for i in range(int(len(string) / 5) + 1):
        print(string[i * 5 : (i + 1) * 5])
        
print_5xn("아이엠어보이유알어걸")

아이엠어보
이유알어걸



### #227. 문자열과 한줄에 출력될 글자 수를 입력을 받아 한 줄에 입력된 글자 수만큼 출력하는 print_mxn(string) 함수를 작성하라.

In [75]:
def print_mxn(string, n):
    for i in range(int(len(string) / n) + 1):
        print(string[i * n : (i + 1) * n])

print_mxn("아이엠어보이유알어걸", 3)

아이엠
어보이
유알어
걸


### #228. 연봉을 입력받아 월급을 계산하는 calc_monthly_salary(annual_salary) 함수를 정의하라. 회사는 연봉을 12개월로 나누어 분할 지급하며, 이 때 1원 미만은 버림한다.

In [76]:
def calc_monthly_salary(annual_salary):
    print(int(annual_salary / 12))
    
calc_monthly_salary(12000000)

1000000


In [77]:
# 정답 - return 사용하기
def calc_monthly_salary(annual_pay) :
    monthly_pay = int(annual_pay / 12)
    return monthly_pay

### #229. 아래 코드의 실행 결과를 예측하라.

In [78]:
def my_print (a, b) :
    print("왼쪽:", a)
    print("오른쪽:", b)

my_print(a=100, b=200)

# 왼쪽: 100
# 오른쪽: 200

왼쪽: 100
오른쪽: 200


### #230. 아래 코드의 실행 결과를 예측하라.

In [79]:
def my_print (a, b) :
    print("왼쪽:", a)
    print("오른쪽:", b)

my_print(b=100, a=200)

# 왼쪽: 200
# 오른쪽: 100

왼쪽: 200
오른쪽: 100


### #231. 아래 코드를 실행한 결과를 예상하라.

In [1]:
def n_plus_1 (n) :
    result = n + 1

n_plus_1(3)
print (result)

# error - 함수 내부에서 사용한 변수는 함수 밖에서 접근이 불가능하므로, 함수 내부에서 계산한 값을 전달하기 위해서는 return을 사용해야 한다.

NameError: name 'result' is not defined

### #232. 문자열 하나를 입력받아 인터넷 주소를 반환하는 make_url 함수를 정의하라.

In [3]:
def make_url(string):
    print("www." + string + ".com")   # return "www." + string + ".com"
    
make_url("naver")

www.naver.com


### #233. 문자열을 입력받아 각 문자들로 구성된 리스트로 반환하는 make_list 함수를 정의하라.

In [7]:
def make_list(string):
    my_list = []
    for i in string:
        my_list.append(i)
    return my_list

make_list("abcd")

['a', 'b', 'c', 'd']

In [8]:
# 리스트 형변환으로 쉽게 풀 수 있다.
def make_list (string) :
    return list(string)

### #234. 숫자로 구성된 하나의 리스트를 입력받아, 짝수들을 추출하여 리스트로 반환하는 pickup_even 함수를 구현하라.

In [9]:
def pickup_even(my_list):
    even_list = []
    for i in my_list:
        if i % 2 == 0:
            even_list.append(i)
    return even_list

pickup_even([3, 4, 5, 6, 7, 8])

[4, 6, 8]

### #235. 콤마가 포함된 문자열 숫자를 입력받아 정수로 변환하는 convert_int 함수를 정의하라.

In [10]:
def convert_int(a):
    return int(a.replace(",", ""))

convert_int("1,234,567")

1234567

### #236. 아래 코드의 실행 결과를 예측하라.

In [11]:
def 함수(num) :
    return num + 4

a = 함수(10)
b = 함수(a)
c = 함수(b)
print(c)

# 22

22


### #237. 아래 코드의 실행 결과를 예측하라.

In [12]:
def 함수(num) :
    return num + 4

c = 함수(함수(함수(10)))
print(c)

# 22

22


### #238. 아래 코드의 실행 결과를 예측하라.

In [13]:
def 함수1(num) :
    return num + 4

def 함수2(num) :
    return num * 10

a = 함수1(10)
c = 함수2(a)
print(c)

# 140

140


### #239. 아래 코드의 실행 결과를 예측하라.

In [14]:
def 함수1(num) :
    return num + 4

def 함수2(num) :
    num = num + 2
    return 함수1(num)

c = 함수2(10)
print(c)

# 16

16


### #240. 아래 코드의 실행 결과를 예측하라.

In [15]:
def 함수0(num) :
    return num * 2

def 함수1(num) :
    return 함수0(num + 2)

def 함수2(num) :
    num = num + 10
    return 함수1(num)

c = 함수2(2)
print(c)

# 28

28


### #241.datetime 모듈을 사용해서 현재 시간을 화면에 출력해보세요.

In [17]:
import datetime

now = datetime.datetime.now()
print(now)

2023-07-28 17:01:10.685553


### #242. datetime 모듈의 now 함수의 리턴 값의 타입을 화면에 출력해보세요.

In [18]:
import datetime

now = datetime.datetime.now()
print(type(now))

<class 'datetime.datetime'>


### #243. datetime 모듈의 timedelta를 사용해서 오늘로부터 5일, 4일, 3일, 2일, 1일 전의 날짜를 화면에 출력해보세요.

In [22]:
import datetime

now = datetime.datetime.now()

for i in range(5, 0, -1):
    other_day = datetime.timedelta(days=i)
    date = now - other_day
    print(date)

2023-07-23 18:55:45.748868
2023-07-24 18:55:45.748868
2023-07-25 18:55:45.748868
2023-07-26 18:55:45.748868
2023-07-27 18:55:45.748868


### #244. 현재시간을 얻어온 후 다음과 같은 포맷으로 시간을 출력해보세요. strftime 메서드를 사용하세요.

> 18:35:01 

In [24]:
import datetime

now = datetime.datetime.now()
time = now.strftime('%H:%M:%S')
print(time)

19:01:32


- [strftime & strptime](https://dev-jy.tistory.com/5)
  - strftime : 날짜와 시간(datetime)을 문자열로 출력
  - strptime : 날짜와 시간 형식의 문자열을 datetime으로 변환

### #245. datetime.datetime.strptime 메서드를 사용하면 문자열 형식의 시간을 datetime.datetime 타입의 시간 값으로 만들어줍니다. "2020-05-04"의 문자열을 시간 타입으로 변환해보세요.

In [26]:
import datetime

day = "2020-05-04"
time_type = datetime.datetime.strptime(day, '%Y-%m-%d')
print(time_type, type(time_type))

2020-05-04 00:00:00 <class 'datetime.datetime'>


### #246. time 모듈, datetime 모듈을 사용해서 1초에 한 번 현재 시간을 출력하는 코드를 작성하세요.

In [30]:
import time
import datetime

while True:
    now = datetime.datetime.now()
    print(now)
    time.sleep(1)

2023-07-28 20:06:05.116552
2023-07-28 20:06:06.121932
2023-07-28 20:06:07.126827
2023-07-28 20:06:08.132134
2023-07-28 20:06:09.137474
2023-07-28 20:06:10.140217
2023-07-28 20:06:11.142528
2023-07-28 20:06:12.145486
2023-07-28 20:06:13.150796
2023-07-28 20:06:14.156063
2023-07-28 20:06:15.159163
2023-07-28 20:06:16.164514
2023-07-28 20:06:17.169295


KeyboardInterrupt: 

- [sleep() 함수](https://blockdmask.tistory.com/571)

### #247. 모듈을 임포트하는 4가지 방식에 대해 설명해보세요.

1. import 모듈
2. import 모듈 as 이름 : 모듈 내의 함수를 호출할 때 모듈 이름 대신 새 이름으로 함수를 호출
3. from 모듈 import 함수명 : 함수를 호출할 때 모듈 이름을 지정하지 않고, 바로 모듈 안의 함수를 호출
4. from 모듈 import * : 모듈 안에 있는 모든 것 (*)을 임포트하는 방식

### #248. os 모듈의 getcwd 함수를 호출하여 현재 디렉터리의 경로를 화면에 출력해보세요.

In [33]:
import os
ret = os.getcwd()
print(ret, type(ret))

/Users/soy/aiffel/afterschool <class 'str'>


### #249. 바탕화면에 텍스트 파일을 하나 생성한 후 os 모듈의 rename 함수를 호출하여 해당 파일의 이름을 변경해보세요.

In [ ]:
#import os
#os.rename("/Users/soy/before.txt", "/Users/soy/after.txt")

### #250. numpy 모듈의 arange 함수를 사용해서 0.0 부터 5.0까지 0.1씩 증가하는 값을 화면에 출력해보세요.

In [35]:
import numpy as np
np.arange(0, 5, 0.1)
# np.arange(시작점(생략 시 0), 끝점(미포함), step size(생략 시 1))

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. , 1.1, 1.2,
       1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2. , 2.1, 2.2, 2.3, 2.4, 2.5,
       2.6, 2.7, 2.8, 2.9, 3. , 3.1, 3.2, 3.3, 3.4, 3.5, 3.6, 3.7, 3.8,
       3.9, 4. , 4.1, 4.2, 4.3, 4.4, 4.5, 4.6, 4.7, 4.8, 4.9])

### #251. 클래스, 객체, 인스턴스에 대해 설명해봅시다.
1. 클래스 : 큰 틀. 객체(object)를 정의하고 만들기 위한 변수와 메서드의 집합.
2. 객체 : 클래스에 의해서 만들어진 물건, 실체. 프로그램에서 구현할 대상. 객체별로 독립적인 성격을 갖는다.
3. 인스턴스 : 프로그램에서 구현된 구체적인 실체. 객체를 프로그램에서 실체화한 것을 인스턴스라고 한다.
- [클래스, 객체, 인스턴스](https://www.itple.co.kr/entry/%ED%8C%8C%EC%9D%B4%EC%8D%AC-%ED%81%B4%EB%9E%98%EC%8A%A4-%EA%B0%9D%EC%B2%B4-%EC%9D%B8%EC%8A%A4%ED%84%B4%EC%8A%A4-%EC%9D%B4%ED%95%B4%ED%95%98%EA%B8%B0)

### #252. 비어있는 사람 (Human) 클래스를 "정의" 해보세요.

In [37]:
class Human:
    pass

### #253. 사람 (Human) 클래스의 인스턴스를 "생성" 하고 이를 areum 변수로 바인딩해보세요.

In [38]:
class Human:
    pass
    
areum = Human()

### #254. 사람 (Human) 클래스에 "응애응애"를 출력하는 생성자를 추가하세요.

In [39]:
class Human:
    def __init__(self):
        print("응애응애")

areum = Human()

응애응애


### #255. 사람 (Human) 클래스에 (이름, 나이, 성별)을 받는 생성자를 추가하세요.

In [43]:
class Human:
    def __init__(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex

### #256. 255에서 생성한 인스턴스의 이름, 나이, 성별을 출력하세요. 인스턴스 변수에 접근하여 값을 출력하면 됩니다.

In [44]:
areum = Human("아름", 25, "여자")
print(areum.name)
print(areum.age)
print(areum.sex)

아름
25
여자


### #257. 사람 (Human) 클래스에서 이름, 나이, 성별을 출력하는 who() 메소드를 추가하세요.

In [45]:
class Human:
    def __init__(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
    
    def who(self):
        print(f'이름: {self.name}, 나이: {self.age}, 성별: {self.sex}')
        
areum = Human("아름", 25, "여자")
areum.who()      # = Human.who(areum)

이름: 아름, 나이: 25, 성별: 여자


### #258. 사람 (Human) 클래스에 (이름, 나이, 성별)을 받는 setInfo 메소드를 추가하세요.

In [46]:
class Human:
    def __init__(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
        
    def who(self):
        print(f'이름: {self.name}, 나이: {self.age}, 성별: {self.sex}')
        
    def setInfo(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
        
areum = Human("불명", "미상", "모름")
areum.who()      # Human.who(areum)

areum.setInfo("아름", 25, "여자")
areum.who()      # Human.who(areum)

이름: 불명, 나이: 미상, 성별: 모름
이름: 아름, 나이: 25, 성별: 여자


### #259. 사람 (human) 클래스에 "나의 죽음을 알리지 말라"를 출력하는 소멸자를 추가하세요.

In [47]:
class Human:
    def __init__(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
        
    def __del__(self):
        print("나의 죽음을 알리지 말라")
        
    def who(self):
        print(f'이름: {self.name}, 나이: {self.age}, 성별: {self.sex}')
        
    def setInfo(self, name, age, sex):
        self.name = name
        self.age = age
        self.sex = sex
        
areum = Human("아름", 25, "여자")
del areum

나의 죽음을 알리지 말라


### #260. 아래와 같은 에러가 발생한 원인에 대해 설명하세요.

In [51]:
class OMG : 
    def print() :
        print("Oh my god")
        
mystock = OMG()
mystock.print()      # OMG.print(mystock)
# OMG class의 print()메소드를 이용하기 위해서는 self로 받는 인자가 없어야 하는데 mystock이라는 인자가 있기 때문이다.

TypeError: OMG.print() takes 0 positional arguments but 1 was given

### #261. 주식 종목에 대한 정보를 저장하는 Stock 클래스를 정의해보세요. 클래스는 속성과 메서드를 갖고 있지 않습니다.

In [52]:
class Stock:
    pass

### #262. Stock 클래스의 객체가 생성될 때 종목명과 종목코드를 입력 받을 수 있도록 생성자를 정의해보세요.

In [54]:
class Stock:
    def __init__(self, name, code):
        self.name = name
        self.code = code
        
삼성 = Stock("삼성전자", "005930")
print(삼성.name)
print(삼성.code)

삼성전자
005930


### #263. 객체에 종목명을 입력할 수 있는 set_name 메서드를 추가해보세요.

In [56]:
class Stock:
    def __init__(self, name, code):
        self.name = name
        self.code = code
        
    def set_name(self, name):
        self.name = name


a = Stock(None, None)
a.set_name("삼성전자")
print(a.name)

삼성전자


### #264. 객체에 종목코드를 입력할 수 있는 set_code 메서드를 추가해보세요.

In [57]:
class Stock:
    def __init__(self, name, code):
        self.name = name
        self.code = code
        
    def set_name(self, name):
        self.name = name
        
    def set_code(self, code):
        self.code = code

a = Stock(None, None)
a.set_code("005930")
print(a.code)

005930


### #265. 종목명과 종목코드를 리턴하는 get_name, get_code 메서드를 추가하세요. 해당 메서드를 사용하여 종목명과 종목코드를 얻고 이를 출력해보세요.

In [60]:
class Stock:
    def __init__(self, name, code):
        self.name = name
        self.code = code
        
    def set_name(self, name):
        self.name = name
        
    def set_code(self, code):
        self.code = code
        
    def get_name(self):
        return self.name
        
    def get_code(self):
        return self.code
    
삼성 = Stock("삼성전자", "005930")
print(삼성.name)
print(삼성.code)
print(삼성.get_name())
print(삼성.get_code())

삼성전자
005930
삼성전자
005930


### #266. 생성자에서 종목명, 종목코드, PER, PBR, 배당수익률을 입력 받을 수 있도록 생성자를 수정하세요. PER, PBR, 배당수익률은 float 타입입니다.

In [63]:
class Stock:
    def __init__(self, name, code, per, pbr, dividend):
        self.name = name
        self.code = code
        self.per = per
        self.pbr = pbr
        self.dividend = dividend
        
    def set_name(self, name):
        self.name = name
        
    def set_code(self, code):
        self.code = code
        
    def get_name(self):
        return self.name
        
    def get_code(self):
        return self.code

### #267. 266번에서 정의한 생성자를 통해 다음 정보를 갖는 객체를 생성해보세요.

> 항목 / 정보  
종목명 / 삼성전자  
종목코드 / 005930  
PER / 15.79  
PBR / 1.33  
배당수익률 / 2.83

In [64]:
삼성 = Stock("삼성전자", "005930", 15.79, 1.33, 2.83)
print(삼성.name)
print(삼성.code)
print(삼성.per)
print(삼성.pbr)
print(삼성.dividend)

삼성전자
005930
15.79
1.33
2.83


### #268. PER, PBR, 배당수익률은 변경될 수 있는 값입니다. 이 값을 변경할 때 사용하는 set_per, set_pbr, set_dividend 메서드를 추가하세요.

In [65]:
class Stock:
    def __init__(self, name, code, per, pbr, dividend):
        self.name = name
        self.code = code
        self.per = per
        self.pbr = pbr
        self.dividend = dividend
        
    def set_name(self, name):
        self.name = name
        
    def set_code(self, code):
        self.code = code
        
    def get_name(self):
        return self.name
        
    def get_code(self):
        return self.code
    
    def set_per(self, per):
        self.per = per
        
    def set_pbr(self, pbr):
        self.pbr = pbr
        
    def set_dividend(self, dividend):
        self.dividend = dividend

### #269. 267번에서 생성한 객체에 set_per 메서드를 호출하여 per 값을 12.75로 수정해보세요.

In [67]:
삼성 = Stock("삼성전자", "005930", 15.79, 1.33, 2.83)
삼성.set_per(12.75)
print(삼성.per)

12.75


### #270. 아래의 표를 참조하여 3종목에 대해 객체를 생성하고 이를 파이썬 리스트에 저장하세요. 파이썬 리스트에 저장된 각 종목에 대해 for 루프를 통해 종목코드와 PER을 출력해보세요.

>종목명 / 종목코드 / PER / PBR / 배당수익률  
삼성전자 / 005930 / 15.79 / 1.33 / 2.83  
현대차 / 005380 / 8.70	/ 0.35 / 4.27  
LG전자 / 066570 / 317.34 / 0.69 / 1.37

In [68]:
종목 = []

삼성 = Stock("삼성전자", "005930", 15.79, 1.33, 2.83)
현대 = Stock("현대차", "005380", 8.70, 0.35, 4.27)
LG = Stock("LG전자", "066570", 317.34, 0.69, 1.37)

종목.append(삼성)
종목.append(현대)
종목.append(LG)

for i in 종목:
    print(i.code, i.per)

005930 15.79
005380 8.7
066570 317.34


### #271. 은행에 가서 계좌를 개설하면 은행이름, 예금주, 계좌번호, 잔액이 설정됩니다. Account 클래스를 생성한 후 생성자를 구현해보세요. 생성자에서는 예금주와 초기 잔액만 입력 받습니다. 은행이름은 SC은행으로 계좌번호는 3자리-2자리-6자리 형태로 랜덤하게 생성됩니다.

In [1]:
import random

class Account:
    def __init__(self, name, balance):
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        
        self.account = num1 + "-" + num2 + "-" + num3
        
kim = Account("김민수", 100)
print(kim.name)
print(kim.balance)
print(kim.bank)
print(kim.account)

김민수
100
sc은행
620-94-671394


- [random() 함수](https://blockdmask.tistory.com/383)   
- [zfill() 함수](https://kkamikoon.tistory.com/entry/Python-%EC%8A%A4%ED%8A%B8%EB%A7%81-%EC%95%9E%EC%97%90-0-%EC%B1%84%EC%9A%B0%EB%8A%94-%EB%B0%A9%EB%B2%95zfill-rjust)

### #272. 클래스 변수를 사용해서 Account 클래스로부터 생성된 계좌 객체의 개수를 저장하세요.

In [1]:
import random

class Account:
    
    count = 0   # class variable
    
    def __init__(self, name, balance):
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
        
kim = Account("김민수", 100)
print(Account.count)
lee = Account("이민수", 100)
print(Account.count)

1
2


### #273. Account 클래스로부터 생성된 계좌의 개수를 출력하는 get_account_num() 메서드를 추가하세요.

In [2]:
import random

class Account:
    
    count = 0    # class variable
    
    def __init__(self, name, balance):
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
        
#    def get_account_num():       # self는 어떤 객체로부터 정보를 가져올지 나타내는 존재.
#        print(Account.count)     # account_num는 어떤 객체로부터 나오는 정보가 아니라 Account class에서 나오는 정보이므로 인자를 없앤다.

    @classmethod                  # 이런 식으로 작성하면 kim.get_account_num()를 사용하더라도 class에 접근한다.
    def get_account_num(cls):
        print(cls.count)

kim = Account("김민수", 100)
lee = Account("이민수", 100)
#Account.get_account_num()
kim.get_account_num()  # Account.get_account_num(kim)

2


### #274. Account 클래스에 입금을 위한 deposit 메서드를 추가하세요. 입금은 최소 1원 이상만 가능합니다.

In [1]:
import random

class Account:
    
    count = 0
    
    def __init__(self, name, balance):
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
        
    @classmethod
    def get_account_num(cls):
        pirnt(cls.count)
        
    def deposit(self, amount):
        if amount >= 1:
            self.balance +=amount

### #275. Account 클래스에 출금을 위한 withdraw 메서드를 추가하세요. 출금은 계좌의 잔고 이상으로 출금할 수는 없습니다.

In [8]:
import random

class Account:
    
    count = 0
    
    def __init__(self, name, balance):
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
    
    @classmethod
    def get_account_num(cls):
        print(cls.count)
        
    def deposit(self, amount):
        if amount >= 1:
            self.balance += amount
        
    def withdraw(self, amount):
        if self.balance > amount:    # balance 앞에 받을 인자를 써주어야 한다.
            self.balance -= amount
        
k = Account("kim", 100)
k.deposit(100)
k.withdraw(90)
print(k.balance)

110


### #276. Account 인스턴스에 저장된 정보를 출력하는 display_info() 메서드를 추가하세요. 잔고는 세자리마다 쉼표를 출력하세요.

In [11]:
import random

class Account:
    
    count = 0
    
    def __init__(self, name, balance):
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
    
    @classmethod
    def get_account_num(cls):
        print(cls.count)
        
    def deposit(self, amount):
        if amount >= 1:
            self.balance += amount
        
    def withdraw(self, amount):
        if self.balance > amount:
            self.balance -= amount
            
    def display_info(self):
        print("은행이름:", self.bank)
        print("예금주:", self.name)
        print("계좌번호:", self.account)
        print("잔고:", f'{self.balance:,}')
        
p = Account("파이썬", 10000)
p.display_info()

은행이름: sc은행
예금주: 파이썬
계좌번호: 266-12-661099
잔고: 10,000


- [숫자에 , 넣는 4가지 방법](https://codechacha.com/ko/python-number-format-comma/)

### #277. 입금 횟수가 5회가 될 때 잔고를 기준으로 1%의 이자가 잔고에 추가되도록 코드를 변경해보세요.

In [12]:
import random

class Account:
    
    count = 0
    
    def __init__(self, name, balance):
        self.deposit_count = 0
        
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
    
    @classmethod
    def get_account_num(cls):
        print(cls.count)
        
    def deposit(self, amount):
        if amount >= 1:
            self.balance += amount
            
        self.deposit_count += 1
        if self.deposit_count % 5 == 0:
            self.balance *= 1.01
        
    def withdraw(self, amount):
        if self.balance > amount:
            self.balance -= amount
            
    def display_info(self):
        print("은행이름:", self.bank)
        print("예금주:", self.name)
        print("계좌번호:", self.account)
        print("잔고:", f'{self.balance:,}')
        
p = Account("파이썬", 10000)
p.deposit(10000)
p.deposit(10000)
p.deposit(10000)
p.deposit(5000)
p.deposit(5000)
print(p.balance)

50500.0


### #278. Account 클래스로부터 3개 이상 인스턴스를 생성하고 생성된 인스턴스를 리스트에 저장해보세요.

In [13]:
import random

class Account:
    
    count = 0
    
    def __init__(self, name, balance):
        self.deposit_count = 0
        
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
    
    @classmethod
    def get_account_num(cls):
        print(cls.count)
        
    def deposit(self, amount):
        if amount >= 1:
            self.balance += amount
            
        self.deposit_count += 1
        if self.deposit_count % 5 == 0:
            self.balance *= 1.01
        
    def withdraw(self, amount):
        if self.balance > amount:
            self.balance -= amount
            
    def display_info(self):
        print("은행이름:", self.bank)
        print("예금주:", self.name)
        print("계좌번호:", self.account)
        print("잔고:", f'{self.balance:,}')
        

data = []        
k = Account("kim", 100000)
l = Account("lee", 1000)
p = Account("Park", 10000)

data.append(k)
data.append(l)
data.append(p)

print(data)

[<__main__.Account object at 0x105c5e980>, <__main__.Account object at 0x105c00160>, <__main__.Account object at 0x105dad120>]


### #279. 반복문을 통해 리스트에 있는 객체를 순회하면서 잔고가 100만원 이상인 고객의 정보만 출력하세요.

In [23]:
import random

class Account:
    
    count = 0
    
    def __init__(self, name, balance):
        self.deposit_count = 0
        
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
    
    @classmethod
    def get_account_num(cls):
        print(cls.count)
        
    def deposit(self, amount):
        if amount >= 1:
            self.balance += amount
            
        self.deposit_count += 1
        if self.deposit_count % 5 == 0:
            self.balance *= 1.01
        
    def withdraw(self, amount):
        if self.balance > amount:
            self.balance -= amount
            
    def display_info(self):
        print("은행이름:", self.bank)
        print("예금주:", self.name)
        print("계좌번호:", self.account)
        print("잔고:", f'{self.balance:,}')
        

data = []        
k = Account("kim", 1000000000)
l = Account("lee", 10000)
p = Account("Park", 1000000)

data.append(k)
data.append(l)
data.append(p)

for i in data:
    if i.balance >= 1000000:
        i.display_info()

은행이름: sc은행
예금주: kim
계좌번호: 238-41-219814
잔고: 1,000,000,000
은행이름: sc은행
예금주: Park
계좌번호: 409-72-362518
잔고: 1,000,000


### #280. 입금과 출금 내역이 기록되도록 코드를 업데이트 하세요. 입금 내역과 출금 내역을 출력하는 deposit_history와 withdraw_history 메서드를 추가하세요.

In [26]:
import random

class Account:
    
    count = 0
    
    def __init__(self, name, balance):
        self.deposit_count = 0
        self.deposit_log = []
        self.withdraw_log = []
        
        self.name = name
        self.balance = balance
        self.bank = "sc은행"
        
        num1 = random.randint(0, 999)
        num2 = random.randint(0, 99)
        num3 = random.randint(0, 999999)
        
        num1 = str(num1).zfill(3)
        num2 = str(num2).zfill(2)
        num3 = str(num3).zfill(6)
        self.account = num1 + "-" + num2 + "-" + num3
        
        Account.count += 1
    
    @classmethod
    def get_account_num(cls):
        print(cls.count)
        
    def deposit(self, amount):
        if amount >= 1:
            self.deposit_log.append(amount)
            self.balance += amount
            
        self.deposit_count += 1
        if self.deposit_count % 5 == 0:
            self.balance *= 1.01
        
    def withdraw(self, amount):
        if self.balance > amount:
            self.withdraw_log.append(amount)
            self.balance -= amount
            
    def display_info(self):
        print("은행이름:", self.bank)
        print("예금주:", self.name)
        print("계좌번호:", self.account)
        print("잔고:", f'{self.balance:,}')
        
    def deposit_history(self):
        for m in self.deposit_log:
            print(m)
    
    def withdraw_history(self):
        for n in self.withdraw_log:
            print(n)
            
k = Account("kim", 1000)
k.deposit(300)
k.deposit(100)
k.withdraw(200)
k.deposit(500)
k.withdraw(100)

k.deposit_history()
print("-----")
k.withdraw_history()

300
100
500
-----
200
100


### #281. 다음 코드가 동작하도록 차 클래스를 정의하세요.

> car = 차(2, 1000)  
> car.바퀴  
-> 2

> car.가격  
-> 1000

In [2]:
class 차:
    def __init__(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격
        
car = 차(2, 1000)
print(car.바퀴)
print(car.가격)

2
1000


### #282. 차 클래스를 상속받은 자전차 클래스를 정의하세요.

In [3]:
class 차:
    def __init__(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격
        
class 자전차(차):
    pass

### #283. 다음 코드가 동작하도록 자전차 클래스를 정의하세요. 단 자전차 클래스는 차 클래스를 상속받습니다.

> bicycle = 자전차(2, 100)  
> bicycle.가격  
-> 100

In [5]:
class 차:
    def __init__(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격
        
class 자전차(차):
    def __init(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격

bicycle = 자전차(2, 100)
print(bicycle.가격)

100


### #284. 다음 코드가 동작하도록 자전차 클래스를 정의하세요. 단 자전차 클래스는 차 클래스를 상속받습니다.

> bicycle = 자전차(2, 100, "시마노")  
> bicycle.구동계  
-> 시마노

In [2]:
class 차:
    def __init__(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격
        
class 자전차(차):
    def __init__(self, 바퀴, 가격, 구동계):
        super().__init__(바퀴, 가격)       #차.__init__(self, 바퀴, 가격)
        self.구동계 = 구동계
        
bicycle = 자전차(2, 100, "시마노")
print(bicycle.구동계)

시마노


### #285. 다음 코드가 동작하도록 차 클래스를 상속받는 자동차 클래스를 정의하세요.

> car = 자동차(4, 1000)  
> car.정보()  
-> 바퀴수 4  
-> 가격 1000

In [3]:
class 차:
    def __init__(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격
        
class 자동차(차):
    def __init(self, 바퀴, 가격):
        super().__init__(바퀴, 가격)
        
    def 정보(self):
        print("바퀴수", self.바퀴)
        print("가격", self.가격)
        
car = 자동차(4, 1000)
car.정보()

바퀴수 4
가격 1000


### #286. 부모 클래스 생성자 호출 - 다음 코드가 동작하도록 차 클래스를 수정하세요.
 
> bicycle = 자전차(2, 100, "시마노")  
> bicycle.정보()    
-> 바퀴수 2  
-> 가격 100

In [4]:
class 차:
    def __init__(self, 바퀴, 가격, 구동계):
        self.바퀴 = 바퀴
        self.가격 = 가격
        self.구동계 = 구동계
        
class 자전차(차):
    def __init__(self, 바퀴, 가격, 구동계):
        super().__init__(바퀴, 가격, 구동계)
    
    def 정보(self):
        print("바퀴수", self.바퀴)
        print("가격", self.가격)
    
bicycle = 자전차(2, 100, "시마노")
bicycle.정보()

바퀴수 2
가격 100


In [1]:
# 정답 - 문제에서 원하는 바를 잘못 파악했다. class 차에서 정보에 대한 메소드를 넣은 후 class 자동차와 class 자전차에서 상속받으면 된다.
class 차:
    def __init__(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격

    def 정보(self):
        print("바퀴수 ", self.바퀴)
        print("가격 ", self.가격)

class 자동차(차):
    def __init__(self, 바퀴, 가격):
        super().__init__(바퀴, 가격)

class 자전차(차):
    def __init__(self, 바퀴, 가격, 구동계):
        super().__init__(바퀴, 가격)
        self.구동계 = 구동계

bicycle = 자전차(2, 100, "시마노")
bicycle.정보()

바퀴수  2
가격  100


### #287. 부모 클래스 메서드 호출 - 자전차의 정보() 메서드로 구동계 정보까지 출력하도록 수정해보세요.
> bicycle = 자전차(2, 100, "시마노")  
> bicycle.정보()  
-> 바퀴수 2  
-> 가격 100  
-> 구동계 시마노

In [3]:
class 차:
    def __init__(self, 바퀴, 가격):
        self.바퀴 = 바퀴
        self.가격 = 가격

    def 정보(self):
        print("바퀴수 ", self.바퀴)
        print("가격 ", self.가격)

class 자동차(차):
    def __init__(self, 바퀴, 가격):
        super().__init__(바퀴, 가격)

class 자전차(차):
    def __init__(self, 바퀴, 가격, 구동계):
        super().__init__(바퀴, 가격)
        self.구동계 = 구동계
        
    def 정보(self):
        super().정보()
        print("구동계", self.구동계)
        
bicycle = 자전차(2, 100, "시마노")
bicycle.정보()

바퀴수  2
가격  100
구동계 시마노


### #288. 메서드 오버라이딩 - 다음 코드의 실행 결과를 예상해보세요.

In [5]:
class 부모:
  def 호출(self):
    print("부모호출")

class 자식(부모):
  def 호출(self):
    print("자식호출")
    
나 = 자식()
나.호출()

# 자식호출 -> super().호출() 과정이 없었기 때문에 부모 클래스의 호출 메소드가 덮어지고 자식의 호출 메소드가 적용이 되어 "자식호출"만 출력된다.

자식호출


### #289. 생성자 - 다음 코드의 실행 결과를 예상해보세요.

In [6]:
class 부모:
  def __init__(self):
    print("부모생성")

class 자식(부모):
  def __init__(self):
    print("자식생성")
    
나 = 자식()
# 자식생성 -> super().__init__()로 부모 클래스의 생성자를 받은 것이 아니기 때문에 자식 클래스의 생성자만 적용이 되어 "자식생성"이 출력된다.

자식생성


### #290. 부모클래스 생성자 호출 - 다음 코드의 실행 결과를 예상해보세요.

In [7]:
class 부모:
  def __init__(self):
    print("부모생성")

class 자식(부모):
  def __init__(self):
    print("자식생성")
    super().__init__()

나 = 자식()
# 자식생성
# 부모생성 -> super().__init__()로 부모 클래스의 생성자도 받았기 때문에 출력 순서대로 자식생성 뒤에 부모생성이 출력된다.

자식생성
부모생성


### #291. 파일 쓰기 - 바탕화면에 '매수종목1.txt' 파일을 생성한 후 다음과 같이 종목코드를 파일에 써보세요.
> 005930  
005380  
035420  

In [30]:
f = open("/Users/soy/aiffel/매수종목1.txt", mode="wt", encoding="utf-8")
f.write("005930\n")
f.write("005380\n")
f.write("035420")
f.close()

### #292. 292 파일 쓰기 - 바탕화면에 '매수종목2.txt' 파일을 생성한 후 다음과 같이 종목코드와 종목명을 파일에 써보세요.
> 005930 삼성전자  
005380 현대차  
035420 NAVER

In [31]:
f = open("/Users/soy/aiffel/매수종목2.txt", mode="wt", encoding="utf-8")
f.write("005930 삼성전자\n")
f.write("005380 현대차\n")
f.write("035420 NAVER")
f.close()

### #293. CSV 파일 쓰기 - 바탕화면에 '매수종목.csv' 파일을 생성한 후 다음과 같이 종목코드와 종목명을 파일에 써보세요. 인코딩은 'cp949'를 사용해야합니다.
![표](https://wikidocs.net/images/page/7044/ch13_csv.PNG)

In [29]:
import csv

f = open("/Users/soy/aiffel/매수종목.csv", mode="wt", encoding="cp949", newline='')
writer = csv.writer(f)
writer.writerow(["종목명", "종목코드","PER"])
writer.writerow(["삼성전자", "005930", 15.79])
writer.writerow(["NAVER", "035420", 55.82])
f.close()

### #294. 파일 읽기 - 바탕화면에 생성한 '매수종목1.txt' 파일을 읽은 후 종목코드를 리스트에 저장해보세요.
>005930  
005380  
035420

In [32]:
f = open("/Users/soy/aiffel/매수종목1.txt", encoding="utf-8")
lines = f.readlines()   # python list

codes = []
for l in lines:
    code = l.strip()    # \n 제거
    codes.append(code)
    
print(codes)

f.close()

['005930', '005380', '035420']


### #295. 파일 읽기 - 바탕화면에 생성한 '매수종목2.txt' 파일을 읽은 후 종목코드와 종목명을 딕셔너리로 저장해보세요. 종목명을 key로 종목명을 value로 저장합니다.

In [35]:
f = open("/Users/soy/aiffel/매수종목2.txt", encoding="utf-8")
lines = f.readlines()    # python list

data = {}
for l in lines:
    l = l.strip()        # \n 제거
    k, v = l.split()
    #print(k, v)
    data[k] = v
    
print(data)
f.close()

{'005930': '삼성전자', '005380': '현대차', '035420': 'NAVER'}


### #296. 예외처리 - 문자열 PER (Price to Earning Ratio) 값을 실수로 변환할 때 에러가 발생합니다. 예외처리를 통해 에러가 발생하는 PER은 0으로 출력하세요.

In [36]:
per = ["10.31", "", "8.00"]

for i in per:
    print(float(i))

10.31


ValueError: could not convert string to float: ''

In [37]:
per = ["10.31", "", "8.00"]

for i in per:
    try:
        print(float(i))
    except:
        print(0)

10.31
0
8.0


### #297 예외처리 및 리스트에 저장 - 문자열로 표현된 PER 값을 실수로 변환한 후 이를 새로운 리스트에 저장해보세요.

In [39]:
per = ["10.31", "", "8.00"]

per_f = []
for i in per:
    try:
        per_f.append(float(i))
    except:
        per_f.append(0)
        
print(per_f)

[10.31, 0, 8.0]


### #298. 특정 예외만 처리하기 - 어떤 값을 0으로 나누면 ZeroDivisionError 에러가 발생합니다. try ~ except로 모든 에러에 대해 예외처리하지 말고 ZeroDivisionError 에러만 예외처리해보세요.

In [46]:
try:
    a = float("")
#    b = 3 / 0
except ZeroDivisionError:
    print("0으로 나눌 수 없어요")

ValueError: could not convert string to float: ''

In [47]:
try:
#    a = float("")
    b = 3 / 0
except ZeroDivisionError:     # ZeroDivisionError만 처리.
    print("0으로 나눌 수 없어요")
    
# 모든 에러를 처리하고 싶으면 except: 사용. 또 다른 에러를 처리하고 싶으면 뒤에 except ~Error:를 쓰면 된다.

0으로 나눌 수 없어요


### #299. 예외의 메시지 출력하기 - 다음과 같은 코드 구조를 사용하면 예외 발생 시 에러 메시지를 변수로 바인딩할 수 있습니다.

> try:  
    실행코드  
except 예외 as 변수:  
    예외처리코드 
### 리스트의 인덱싱에 대해 에러를 출력해보세요.

In [48]:
data = [1, 2, 3]

for i in range(5):
    try:
        print(data[i])
    except IndexError as e:       # IndexError를 e라는 변수에 바인딩
#        print("IndexError 발생")   # 다른 문구도 출력 가능 
        print(e)

1
2
3
list index out of range
list index out of range


### #300. try, except, else, finally 구조 사용해보기 - 파이썬 예외처리는 다음과 같은 구조를 가질 수 있습니다.

>try:  
    실행 코드  
except:  
    예외가 발생했을 때 수행할 코드  
else:  
    예외가 발생하지 않았을 때 수행할 코드  
finally:  
    예외 발생 여부와 상관없이 항상 수행할 코드  
### 아래의 코드에 대해서 예외처리를 사용하고 try, except, else, finally에 적당한 코드를 작성해봅시다. else와 finally는 적당한 문구를 print하시면 됩니다.

In [45]:
per = ["10.31", "", "8.00"]

for i in per:
    try:
        print(float(i))
    except:
        print(0)
    else:
        print("claen")
    finally:
        print("화이팅!!")

10.31
claen
화이팅!!
0
화이팅!!
8.0
claen
화이팅!!
